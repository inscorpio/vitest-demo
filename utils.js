export const getRepeatItems = (input) => {
  const valueMap = input.reduce((res, item) => {
    const items = res.get(item.num1) ?? []
    res.set(item.num1, [...items, item])
    return res
  }, new Map())

  const valueGroups = Array.from(valueMap).map((v) => v[1])
  return valueGroups.reduce((res, groupItems) => {
    const symbolMap = new Map()
    groupItems.forEach((item) => {
      const symbols = item.symbol.split('')
      symbols.forEach((symbol) => {
        const items = symbolMap.get(symbol) ?? []
        symbolMap.set(symbol, [...items, item])
      })
    })
    const symbolGroups = Array.from(symbolMap).map(v => v[1]).filter((v) => v.length > 1)

    res.push(...symbolGroups.flat().slice(1))
    return res
  }, [])
}
