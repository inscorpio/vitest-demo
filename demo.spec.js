import { it, expect } from 'vitest'
import { getRepeatItems } from './utils'

it('should return repeat item', () => {
  const input = [
    {
      symbol: '=',
      num1: '0',
    },
    {
      symbol: '>=',
      num1: '0',
    },
    {
      symbol: '>=',
      num1: '5',
    },
    {
      symbol: '<=',
      num1: '5',
    },
    {
      symbol: '=',
      num1: '99',
    },
    {
      symbol: '=',
      num1: '99',
    },
    {
      symbol: '>',
      num1: '10',
    },
    {
      symbol: '>=',
      num1: '10',
    },
    {
      symbol: '>',
      num1: '10',
    },
  ]
  expect(getRepeatItems(input)).toMatchSnapshot()
})
