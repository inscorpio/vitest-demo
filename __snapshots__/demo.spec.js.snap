// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return repeat item 1`] = `
[
  {
    "num1": "0",
    "symbol": ">=",
  },
  {
    "num1": "5",
    "symbol": "<=",
  },
  {
    "num1": "10",
    "symbol": ">=",
  },
  {
    "num1": "10",
    "symbol": ">",
  },
  {
    "num1": "99",
    "symbol": "=",
  },
]
`;
